From bb201ca3e69bd3366665cb6053efb31b2082583f Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 13 Nov 2018 22:45:06 +0100
Subject: [PATCH 2/2] pie-gsi tracking

commit 2c8f003f20c26d36055f72b913d82deca5ff733e
Merge: 24e119892 e020e3baf
Author: Treehugger Robot <treehugger-gerrit@google.com>
Date:   Tue Oct 30 05:40:55 2018 +0000

    Merge changes I32af5459,I198fb9b7,I0d61ea11,I59d9b8f7 into pie-gsi

    * changes:
      Unzip bin and libs from otatools.zip for build_mixed
      Revise comments for when -v, -m options are needed
      Check first if system_dir and device_dir exists
      Print error on missing device target_files archive

commit e020e3bafc86bf6587104f12b1bafd7acf8c8196
Author: Justin Yun <justinyun@google.com>
Date:   Fri Oct 26 12:21:24 2018 +0900

    Unzip bin and libs from otatools.zip for build_mixed

    For build_mixed, only "bin" and "lib64" are required from the
    otatools.zip.
    Instead of unzipping all files, unzip only the required files from
    the otatools.zip.

    Also update the usage for -t option.

    Bug: 112732863
    Test: build_mixed -v 9 -m modify_system_img.sh -p vbmeta.img \
                      -t otatools.zip system product dist checkvintf

    Change-Id: I32af5459f6872f15cda01628387357b8e7492609
    (cherry picked from commit ec96edb752253bbb01aa15463d0524621d04ad25)

commit fc7d769214b690310dbaaec646bd9a25905a9613
Author: Jae Shin <jaeshin@google.com>
Date:   Tue Aug 14 14:20:12 2018 +0900

    Revise comments for when -v, -m options are needed

    Test: build_mixed
    Change-Id: I198fb9b7d9916ee7bb3ab5e005a3de1e4d1dd866
    (cherry picked from commit d5fd24f609345d63e29f97771ff5a6023731df29)

commit 7ebc00199cfbd9f9671f278ae499afafd2aa41d3
Author: Jae Shin <jaeshin@google.com>
Date:   Thu Aug 9 12:57:20 2018 +0900

    Check first if system_dir and device_dir exists

    Change the order of variable checking so it is clear whether
    the system_dir or device_dir does not exist or a specific file
    in the *_dir does not exist.

    Test: build_mixed system_dir device_dir dist_dir
    Merged-In: I0d61ea11c3bd22ef12eccbb46b8f665414188307
    Change-Id: I0d61ea11c3bd22ef12eccbb46b8f665414188307
    (cherry picked from commit 52edf8fb4f5eaa8f6c6c13a1a87a34311ccb76ed)
    (cherry picked from commit f418b9b9e55fc1d4e658a2942ed7bb6ce1972836)

commit 66118192e209f4bd3105889a63ff722f28ea0c69
Author: Paul Trautrim <paultrautrim@google.com>
Date:   Fri Jun 29 16:19:10 2018 +0900

    Print error on missing device target_files archive

    Check the newly-created DEVICE_TARGET_FILES_ARCHIVE variable instead of
    checking the DEVICE_ARCHIVE variable twice.

    Test: run build_mixed with only the *-img-* archive in the device directory
    Merged-In: I59d9b8f77cac1dcdeb2c88cdf37bc8dfbd1111ca
    Change-Id: I59d9b8f77cac1dcdeb2c88cdf37bc8dfbd1111ca
    (cherry picked from commit c1474312b3babf7470508382a41011aaa94a98b7)
    (cherry picked from commit d2d8ac950940bf8ceca5d1ea3cdb4ed162510d8a)

commit 24e119892b223e580512da988d5c8eb0a6f653f1
Merge: 924c2fde2 1722a9515
Author: Treehugger Robot <treehugger-gerrit@google.com>
Date:   Mon Oct 29 06:50:51 2018 +0000

    Merge "Use prebuilt otatools for mixed build" into pie-gsi

commit 1722a951588b44bc14114807edd72100c9c90782
Author: Justin Yun <justinyun@google.com>
Date:   Fri Oct 5 14:54:17 2018 +0900

    Use prebuilt otatools for mixed build

    Enable mixed_build script to use prebuilt otatools for packaging
    image files. If we have otatools.zip file, use '-t' option to use
    the prebuilt tools instead of building them in mixed branches.

    Bug: 112732863
    Test: run mixed_build with otatools using the command below:
     $ development/vndk/tools/build_mixed -v 9 \
       -m development/vndk/tools/modify_system_img.sh \
       -p out/prebuilt_cached/artifacts/gsi/vbmeta.img \
       -t out/prebuilt_cached/artifacts/gsi/otatools.zip \
       out/prebuilt_cached/artifacts/gsi \
       out/prebuilt_cached/artifacts/taimen \
       dist \
       out/prebuilt_cached/artifacts/gsi/checkvintf

    Change-Id: I7891a2bad2f4e471baffb987149a39bbbd57f26e
    (cherry picked from commit 40da678a9fd22dae2ac37c6945ca09afd31e494d)

commit 924c2fde2dc4e3709376d877197730666652bf63
Merge: 60ec2abfe 656ef23ba
Author: Treehugger Robot <treehugger-gerrit@google.com>
Date:   Mon Sep 10 11:15:01 2018 +0000

    Merge changes from topic "from-pi-gsi" into pie-gsi

    * changes:
      Revert "Move SdkSetup.apk to vendor partition"
      Move SdkSetup.apk to vendor partition

commit 656ef23ba965d8a5f0116af3b34defb74ea8ccda
Author: Antoan Angelov <arangelov@google.com>
Date:   Thu Jul 19 12:57:33 2018 +0000

    Revert "Move SdkSetup.apk to vendor partition"

    This reverts commit 3bf31401fe89b1941132d2c9159e290dfab3b2df.

    Reason for revert: Build cop revert because this change is breaking builds.

    Change-Id: I54f87f21eec0e93873ffedec6b0bbe154584ba4b
    Merged-Id: I54f87f21eec0e93873ffedec6b0bbe154584ba4b
    (cherry picked from commit e867e27a3b4ef12155d7c064a6816157954d172d)

commit c5e7ae8b0a39fd434c014e8bb012506308018709
Author: Bowgo Tsai <bowgotsai@google.com>
Date:   Tue Jul 17 18:06:17 2018 +0800

    Move SdkSetup.apk to vendor partition

    Otherwise, it causes the CTS-on-GSI failure on non-emulator devices:
      CtsPermission2TestCases android.permission2.cts.PrivappPermissionsTest#testPrivappPermissionsEnforcement

    Because non-emulator devices lack the permission file in their /vendor.
      /system/priv-app/SdkSetup/SdkSetup.apk
      /vendor/etc/permissions/privapp-permissions-goldfish.xml

    Bug: 110517983
    Test: boot a GSI and checks that SdkSetup.apk is moved to /vendor
    Change-Id: I3086f99c477eeca38f42041e1809a5aa3b8bfa7a
    Merged-In: I3086f99c477eeca38f42041e1809a5aa3b8bfa7a
    (cherry picked from commit 3bf31401fe89b1941132d2c9159e290dfab3b2df)

Change-Id: Ia51887a0ce195176692dc2776cf48776e5fc6c25
---
 vndk/tools/build_mixed | 46 ++++++++++++++++++++++++++++--------------
 1 file changed, 31 insertions(+), 15 deletions(-)

diff --git a/vndk/tools/build_mixed b/vndk/tools/build_mixed
index 5ff36d40f..cb4055d86 100755
--- a/vndk/tools/build_mixed
+++ b/vndk/tools/build_mixed
@@ -2,16 +2,19 @@
 usage () {
     echo "Create a Mixed Build archive with the given system and device archives."
     echo
-    echo "Usage: $0 [-v <vendor_version>] [-m <modify_system_image_path>] [-p <override_vbmeta_image_path>]"
+    echo "Usage: $0 [-v <vendor_version>] [-m <modify_system_image_path>]"
+    echo "    [-t <prebuilt_otatools_path>] [-p <override_vbmeta_image_path>]"
     echo "    system_build_dir device_build_dir out_dir [check_tool]"
     echo
-    echo "Options -v, -m, -p must precede positional arguments."
+    echo "Options -v, -m, -t, -p must precede positional arguments."
     echo
-    echo "vendor_version is the version of the vendor image when the mixed"
-    echo "    build is inter branch. Optional."
+    echo "vendor_version is the version of the vendor image when Keymaster v3"
+    echo "    related modifications to the system image is necessary. Optional."
     echo "    eg. 8.1.0 for a mixed build of GSI and O-MR1 vendor image."
     echo "modify_system_image_path is the path to the script that modifies the"
-    echo "    system image for an inter branch build target. Optional."
+    echo "    system image, needed for Keymaster v3. Optional."
+    echo "prebuilt_otatools_path is the path to otatools.zip file that has all"
+    echo "    required host binaries to modify system image. Optional."
     echo "override_vbmeta_image_path is the path to a vbmeta.img to use"
     echo "    to override the existing vbmeta.img of device. Optional."
     echo "system_build_dir is the path to the system build"
@@ -41,7 +44,7 @@ cleanup_and_exit () {
 
 trap cleanup_and_exit EXIT
 
-while getopts :v:m:p: opt; do
+while getopts :v:m:p:t: opt; do
     case "$opt" in
         v)
             readonly VENDOR_VERSION="$OPTARG"
@@ -52,6 +55,9 @@ while getopts :v:m:p: opt; do
         p)
             readonly OVERRIDE_VBMETA_IMAGE_PATH="$OPTARG"
             ;;
+        t)
+            readonly OTATOOLS_ZIP="$OPTARG"
+            ;;
         \?)
             exit_badparam "Invalid options: -"$OPTARG""
             ;;
@@ -67,14 +73,6 @@ fi
 
 shift "$((OPTIND-1))"
 
-if [[ ! -z "${MODIFY_SYSTEM_SCRIPT+x}" && ! -f "$MODIFY_SYSTEM_SCRIPT" ]]; then
-    exit_badparam "Script not found: "$MODIFY_SYSTEM_SCRIPT""
-fi
-
-if [[ ! -z "${OVERRIDE_VBMETA_IMAGE_PATH+x}" && ! -f "$OVERRIDE_VBMETA_IMAGE_PATH" ]]; then
-    exit_badparam "Specified vbmeta.img not found: "$OVERRIDE_VBMETA_IMAGE_PATH""
-fi
-
 if [[ $# -lt 3 ]]; then
     exit_badparam "Unexpected number of arguments"
 fi
@@ -96,14 +94,23 @@ if [[ ! -f "$DEVICE_ARCHIVE" ]]; then
 fi
 
 readonly DEVICE_TARGET_FILES_ARCHIVE="$(find "$DEVICE_DIR" -name "*-target_files-*.zip" -print)"
-if [[ ! -f "$DEVICE_ARCHIVE" ]]; then
+if [[ ! -f "$DEVICE_TARGET_FILES_ARCHIVE" ]]; then
     exit_badparam "Could not find device target_files archive in $DEVICE_DIR."
 fi
 
+if [[ ! -z "${MODIFY_SYSTEM_SCRIPT+x}" && ! -f "$MODIFY_SYSTEM_SCRIPT" ]]; then
+    exit_badparam "Script not found: "$MODIFY_SYSTEM_SCRIPT""
+fi
+
+if [[ ! -z "${OVERRIDE_VBMETA_IMAGE_PATH+x}" && ! -f "$OVERRIDE_VBMETA_IMAGE_PATH" ]]; then
+    exit_badparam "Specified vbmeta.img not found: "$OVERRIDE_VBMETA_IMAGE_PATH""
+fi
+
 readonly DEVICE_ARTIFACTS_DIR="$TEMP_DIR"/device_archive_artifacts
 readonly DEVICE_IMAGES_DIR="$DEVICE_ARTIFACTS_DIR"/IMAGES
 readonly SYSTEM_ARTIFACTS_DIR="$TEMP_DIR"/system_artifacts
 readonly SYSTEM_IMAGES_DIR="$SYSTEM_ARTIFACTS_DIR"/IMAGES
+readonly OTATOOLS_DIR="$TEMP_DIR"/otatools
 
 readonly SPL_PROPERTY_NAME="ro.build.version.security_patch"
 readonly SYSTEM_BUILD_PROP="SYSTEM/build.prop"
@@ -124,6 +131,15 @@ unzip "$DEVICE_TARGET_FILES_ARCHIVE" \
   META/vendor_matrix.xml META/vendor_manifest.xml "$SYSTEM_BUILD_PROP" \
   -d "$DEVICE_ARTIFACTS_DIR"
 
+if [[ -f "$OTATOOLS_ZIP" ]]; then
+    # Uncompress otatools
+    mkdir -p "$OTATOOLS_DIR"
+    unzip "$OTATOOLS_ZIP" bin/* lib64/* -d "$OTATOOLS_DIR"
+    # Set paths for using prebuilt host binaries.
+    export PATH="$OTATOOLS_DIR"/bin:"$PATH"
+    export LD_LIBRARY_PATH="$OTATOOLS_DIR"/lib64:"$LD_LIBRARY_PATH"
+fi
+
 ###
 # Check compatibility between the system and device.
 if [[ -f "$CHECK_TOOL" ]]; then
-- 
2.17.1

